core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala

```diff
@@ -4342,6 +4342,56 @@ class DAGSchedulerSuite extends SparkFunSuite with TempLocalSparkContext with Ti
     assertDataStructuresEmpty()
   }
 
+  test("SPARK-38987: corrupted shuffle block FetchFailure should unregister merge result") {
+    initPushBasedShuffleConfs(conf)
+    DAGSchedulerSuite.clearMergerLocs()
+    DAGSchedulerSuite.addMergerLocs(Seq("host1", "host2", "host3", "host4", "host5"))
+
+    scheduler = new MyDAGScheduler(
+      sc,
+      taskScheduler,
+      sc.listenerBus,
+      mapOutputTracker,
+      blockManagerMaster,
+      sc.env,
+      shuffleMergeFinalize = false,
+      shuffleMergeRegister = false)
+    dagEventProcessLoopTester = new DAGSchedulerEventProcessLoopTester(scheduler)
+
+    val parts = 2
+    val shuffleMapRdd = new MyRDD(sc, parts, Nil)
+    val shuffleDep = new ShuffleDependency(shuffleMapRdd, new HashPartitioner(parts))
+    val reduceRdd = new MyRDD(sc, parts, List(shuffleDep), tracker = mapOutputTracker)
+
+    // Submit a reduce job that depends which will create a map stage
+    submit(reduceRdd, (0 until parts).toArray)
+
+    // Pass in custom bitmap so that the mergeStatus can store
+    // the correct mapIndex.
+    val bitmap = new RoaringBitmap()
+    bitmap.add(-1)
+
+    val shuffleMapStage = scheduler.stageIdToStage(0).asInstanceOf[ShuffleMapStage]
+    scheduler.handleRegisterMergeStatuses(shuffleMapStage,
+      Seq((0, MergeStatus(makeBlockManagerId("hostA"), shuffleDep.shuffleMergeId, bitmap, 1000L))))
+    scheduler.handleShuffleMergeFinalized(shuffleMapStage,
+      shuffleMapStage.shuffleDep.shuffleMergeId)
+    scheduler.handleRegisterMergeStatuses(shuffleMapStage,
+      Seq((1, MergeStatus(makeBlockManagerId("hostA"), shuffleDep.shuffleMergeId, bitmap, 1000L))))
+
```

Comment:
We can remove `bitmap`, use `makeMergeStatus` to simplify the test.
You can refer to existing tests for push based shuffle for this.

Reply:
I tried using that, but `makeMergeStatus` uses a mock bitmap - so the mapIndex is never present inside the `MergeStatus`. Hence, I opted to use the actual class, and it seemed like this was the convention being used for triggering `MapOutputTracker.unregisterMergeResult` inside `MapOutputTrackerSuite`.

Reply:
Let me know if i am missing something

Reply:
You are right, we need the bitmap to be valid, and not mocked.